Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbQueryCommandBuilder.cs

```diff
@@ -0,0 +1,104 @@
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+using System.Text;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.Model;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.Parameters;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer.Parameters;
+using Remotion.Data.DomainObjects.Queries;
+
+namespace Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer.DbCommandBuilders;
+
+public class SqlDbQueryCommandBuilder : QueryDbCommandBuilder
+{
+  private readonly IReadOnlyDictionary<string,TvpSpec> _tvpToTempTable;
+
+  private struct TvpSpec
+  {
+    public TvpSpec (string tempTableName, string indexName, bool hasIndex)
+    {
+      TempTableName = tempTableName;
+      IndexName = indexName;
+      HasIndex = hasIndex;
+    }
+
+    public string IndexName { get; }
+    public string TempTableName { get; }
+    public bool HasIndex { get; }
+  }
+
+  public SqlDbQueryCommandBuilder (string statement, IEnumerable<QueryParameterWithDataParameterDefinition> parameters, ISqlDialect sqlDialect)
+      : base(statement, parameters, sqlDialect)
+  {
+    var tvp = ParametersWithDefinition.Where(p => p.DataParameterDefinition is SqlTableValuedDataParameterDefinition);
+    _tvpToTempTable = tvp.ToDictionary(p => p.QueryParameter.Name, p => new TvpSpec(
+        $"[#{p.QueryParameter.Name}]",
+        $"[IX_{p.QueryParameter.Name}]",
+        ShouldCreateIndex(((SqlTableValuedDataParameterDefinition)p.DataParameterDefinition).RecordDefinition)));
+  }
+
+  private static bool ShouldCreateIndex (RecordDefinition recordDefinition)
+  {
+    if (recordDefinition.PropertyDefinitions.Count > 1)
+      return false;
+
+    var storagePropertyDefinition = recordDefinition.PropertyDefinitions.Single().StoragePropertyDefinition;
+    if (storagePropertyDefinition is SerializedObjectIDStoragePropertyDefinition)
+      return true;
+
+    if (storagePropertyDefinition is not SimpleStoragePropertyDefinition simpleStoragePropertyDefinition)
+      return false;
+
+    var storageTypeLength = simpleStoragePropertyDefinition.ColumnDefinition.StorageTypeInfo.StorageTypeLength;
+    if(!storageTypeLength.HasValue)
+      return true;
+
+    if (storageTypeLength.Value < 0)
+      return false;
+
+    var storageDbType = simpleStoragePropertyDefinition.ColumnDefinition.StorageTypeInfo.StorageDbType;
+    var maxLength = storageDbType == DbType.String || storageDbType == DbType.StringFixedLength
+        ? 1700 / 2
+        : 1700;
+
+    return storageTypeLength <= maxLength;
+  }
+
+  public override IDbCommand Create (IDbCommandFactory dbCommandFactory)
+  {
+    var command = base.Create(dbCommandFactory);
+
+    var commandTextBuilder = new StringBuilder(command.CommandText.Length);
+
+    foreach (var tvp in _tvpToTempTable)
+    {
+      commandTextBuilder.AppendLine($"SELECT [Value] INTO {tvp.Value.TempTableName} FROM {tvp.Key};");
```

Comment:
we use SqlDialect.DelimitIdentifier also in SqlTableValuedParameterComparedColumnSpecification. I can't say if that's just the only one where it's done that way or if this is the only possible place anyway. It just feels strange not doing it, but that's just my gut feeling